anndata/dev/multi_files/_anndataset.py

```diff
@@ -0,0 +1,670 @@
+            layers = None
+        else:
+            layers = self.layers.to_dict(use_convert=False)
+        obsm = None if self.obsm is None else self.obsm.to_dict(use_convert=False)
+        obs = (
+            None
+            if self.obs is None
+            else pd.DataFrame(self.obs.to_dict(use_convert=False))
+        )
+
+        if ignore_X:
+            X = None
+            shape = self.shape
+        else:
+            X = self._gather_X()
+            shape = None
+
+        adata = AnnData(X, obs=obs, obsm=obsm, layers=layers, shape=shape)
+        adata.obs_names = self.obs_names
+        adata.var_names = self.var_names
+        return adata
+
+
+class AnnDataSet(_ConcatViewMixin):
+    def __init__(
+        self,
+        adatas,
+        join_obs="inner",
+        join_obsm=None,
+        join_vars=None,
+        label=None,
+        keys=None,
+        index_unique=None,
+        convert=None,
+        harmonize_dtypes=True,
+        indices_strict=True,
+    ):
+        if isinstance(adatas, Mapping):
+            if keys is not None:
+                raise TypeError(
+                    "Cannot specify categories in both mapping keys and using `keys`. "
+                    "Only specify this once."
+                )
+            keys, adatas = list(adatas.keys()), list(adatas.values())
+        else:
+            adatas = list(adatas)
+
+        # check if the variables are the same in all adatas
+        self.adatas_vidx = [None for adata in adatas]
+        vars_names_list = [adata.var_names for adata in adatas]
+        vars_eq = all([adatas[0].var_names.equals(vrs) for vrs in vars_names_list[1:]])
+        if vars_eq:
+            self.var_names = adatas[0].var_names
+        elif join_vars == "inner":
+            var_names = reduce(pd.Index.intersection, vars_names_list)
+            self.adatas_vidx = []
+            for adata in adatas:
+                if var_names.equals(adata.var_names):
+                    self.adatas_vidx.append(None)
+                else:
+                    adata_vidx = _normalize_index(var_names, adata.var_names)
+                    self.adatas_vidx.append(adata_vidx)
+            self.var_names = var_names
+        else:
+            raise ValueError(
+                "Adatas have different variables. "
+                "Please specify join_vars='inner' for intersection."
+            )
+
+        concat_indices = pd.concat(
+            [pd.Series(a.obs_names) for a in adatas], ignore_index=True
+        )
+        if keys is None:
+            keys = np.arange(len(adatas)).astype(str)
+        label_col = pd.Categorical.from_codes(
+            np.repeat(np.arange(len(adatas)), [a.shape[0] for a in adatas]),
+            categories=keys,
+        )
+        if index_unique is not None:
+            concat_indices = concat_indices.str.cat(
+                label_col.map(str), sep=index_unique
+            )
+        self.obs_names = pd.Index(concat_indices)
```

Comment:
Shouldn't this set the index of a `self._obs` attribute that stores the joint obs annotations? And `self.obs` is a property?

A similar comment applies below.

Reply:
Done.